src/include/libpmemobj.h

```diff
@@ -86,6 +101,580 @@ void pmemobj_set_funcs(
 		void *(*malloc_func)(size_t size),
 		void (*free_func)(void *ptr));
 
+/*
+ * Locking...
+ */
+#define	_POBJ_CL_ALIGNMENT 64 /* cache line alignment for performance */
+
+typedef union padded_pmemmutex {
+	char padding[_POBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_mutex_t mutex;
+	} pmemmutex;
+} PMEMmutex;
+
+typedef union padded_pmemrwlock {
+	char padding[_POBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_rwlock_t rwlock;
+	} pmemrwlock;
+} PMEMrwlock;
+
+typedef union padded_pmemcond {
+	char padding[_POBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_cond_t cond;
+	} pmemcond;
+} PMEMcond;
+
+int pmemobj_mutex_zero(PMEMmutex *mutexp);
+int pmemobj_mutex_lock(PMEMobjpool *pop, PMEMmutex *mutexp);
+int pmemobj_mutex_trylock(PMEMobjpool *pop, PMEMmutex *mutexp);
+int pmemobj_mutex_unlock(PMEMobjpool *pop, PMEMmutex *mutexp);
+
+int pmemobj_rwlock_zero(PMEMrwlock *rwlockp);
+int pmemobj_rwlock_rdlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_wrlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_timedrdlock(PMEMobjpool *pop,
+	PMEMrwlock *restrict rwlockp,
+	const struct timespec *restrict abs_timeout);
+int pmemobj_rwlock_timedwrlock(PMEMobjpool *pop,
+	PMEMrwlock *restrict rwlockp,
+	const struct timespec *restrict abs_timeout);
+int pmemobj_rwlock_tryrdlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_trywrlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_unlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+
+int pmemobj_cond_zero(PMEMcond *condp);
+int pmemobj_cond_broadcast(PMEMobjpool *pop, PMEMcond *condp);
+int pmemobj_cond_signal(PMEMobjpool *pop, PMEMcond *condp);
+int pmemobj_cond_timedwait(PMEMobjpool *pop, PMEMcond *restrict condp,
+	PMEMmutex *restrict mutexp, const struct timespec *restrict abstime);
+int pmemobj_cond_wait(PMEMobjpool *pop, PMEMcond *condp,
+	PMEMmutex *restrict mutexp);
+
+/* find a offset of pool identified by uuid_lo */
+uint64_t pmemobj_off_by_uuid_lo(uint64_t uuid_lo);
+
+/*
+ * Persistent memory object
+ */
+
+/*
+ * Object handle
+ */
+typedef struct pmemoid {
+	uint64_t pool_uuid_lo;
+	uint64_t off;
+} PMEMoid;
+
+#define	OID_TYPE(type)\
+union {\
+	type *__type;\
+	PMEMoid oid;\
+}
+
+#define	OID_ASSIGN(o, value) ((o).oid = value)
+
+#define	OID_ASSIGN_TYPED(lhs, rhs)\
+__builtin_choose_expr(\
+	__builtin_types_compatible_p(\
+		typeof((lhs).__type),\
+		typeof((rhs).__type)),\
+	(void) ((lhs).oid = (rhs).oid),\
+	(lhs.__type = rhs.__type))
+
+#define	OID_NULL		((PMEMoid) {0, 0})
+#define	_POBJ_MAX_OID_TYPE_NUM	1024
+
+#define	OID_IS_NULL(o)	((o).oid.off == 0)
+
+#define	OID_EQUALS(lhs, rhs)\
+((lhs).oid.off == (rhs).oid.off &&\
+	(lhs).oid.pool_uuid_lo == (rhs).oid.pool_uuid_lo)
+
+#define	_POBJ_DIRECT_OFF(o)\
```

Comment:
Seems like there is no (easy) way to get a direct pointer from *PMEMoid*. *DIRECT_* macros work with typed OID's only. Maybe (for those who don't care about type-safety) we should have a function like:
```C
void *pmemobj_direct(PMEMoid oid) {
    return (oid.off + pmemobj_off_by_uuid_lo(oid.pool_uuid_lo));
}
```
and simply use it in _POBJ_DIRECT_OFF macro:
```C
#define	_POBJ_DIRECT_OFF(o) pmemobj_direct(o.oid)
```